### PR TITLE
CMK

### PR DESCRIPTION
**Pull request by dagiro**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

**Comment by bmoore-msft** [comment]
@cmk - see: https://github.com/Azure/azure-quickstart-templates/tree/master/1-CONTRIBUTION-GUIDE#azure-devops-ci for how to supply params for a sample

File: 101-hdinsight-spark-cmk/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+                },
+                "minSupportedTlsVersion": "[parameters('minTlsVersionNumber')]",
+                "diskEncryptionProperties": {
+                    "vaultUri": "[parameters('diskEncryptionVaultUri')]",
+                    "keyName": "[parameters('diskEncryptionKeyName')]",
+                    "keyVersion": "[parameters('diskEncryptionKeyVersion')]",
+                    "msiResourceId": "[parameters('diskEncryptionMsiResourceId')]"
+                }
+            },
+            "identity": {
```
Comment (karunasagark) [inline 101-hdinsight-spark-cmk/azuredeploy.json]: This "identity" would not work here. It needs to be placed as a sibling of "properties".

Comment (dagiro) [reply 101-hdinsight-spark-cmk/azuredeploy.json]: It is under properties

Comment (karunasagark) [reply 101-hdinsight-spark-cmk/azuredeploy.json]: It should be a "sibling", not "under". And please make sure you try and this works.

Comment (dagiro) [reply 101-hdinsight-spark-cmk/azuredeploy.json]: What error did you receive when you executed this?

**Comment by dagiro** [comment]
DeploymentName          : Spark_CKM
ResourceGroupName       : v-dagiro_RG_East
ProvisioningState       : Succeeded
Timestamp               : 3/26/2020 10:12:52 PM
Mode                    : Incremental
TemplateLink            : 
Parameters              :

**Comment by dagiro** [comment]
@bmoore-msft changes made

**Comment by SnehaGunda** [comment]
#sign-off

**Comment by SnehaGunda** [comment]
@dagiro the checks failed again this tile it just couldn't deploy the template. After removing the empty properties, did you deploy the template and check if it was working.

**Comment by dagiro** [comment]
@bmoore-msft, regarding the empty value best practice error for line 185, I can't get a deployment to work **without** the space.
Error: Code=InvalidTemplate; Message=Deployment template parse failed: 'Error converting value 
"/subscriptions/e74f0ec1-631b-46f4-bde5-6aa999cf5be0/resourcegroups/v-dagiro_RG_East/providers/Microsoft.ManagedIdentity/userAssignedIdentities/GirouxUAMI" to type 
'Microsoft.WindowsAzure.ResourceStack.Common.Collections.InsensitiveDictionary`1[Newtonsoft.Json.Linq.JToken]'. Path 'userAssignedIdentities'.'.

**Comment by bmoore-msft** [comment]
Don't worry about the empty property check - we'll update the test to cover that scenario.

**Comment by omidm1** [comment]
#sign-off

File: 101-hdinsight-spark-cmk/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  "parameters": {
+    "clusterName": {
+      "type": "string",
+      "metadata": {
+        "description": "The name of the HDInsight cluster to create."
+      }
+    },
+    "clusterLoginUserName": {
+      "type": "string",
+      "defaultValue": "admin",
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.json]: remove defaultValue

File: 101-hdinsight-spark-cmk/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    "clusterWorkerNodeCount": {
+      "type": "int",
+      "defaultValue": 1,
+      "metadata": {
+        "description": "The number of nodes in the HDInsight cluster."
+      }
+    },
+    "clusterKind": {
+      "type": "string",
+      "defaultValue": "SPARK",
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.json]: is there a set of allowed values we should provide here?

File: 101-hdinsight-spark-cmk/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      "metadata": {
+        "description": "The type of the HDInsight cluster to create."
+      }
+    },
+    "sshUserName": {
+      "type": "string",
+      "defaultValue": "sshuser",
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.json]: remove defaultValue

File: 101-hdinsight-spark-cmk/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        "description": "These credentials can be used to remotely access the cluster."
+      }
+    },
+    "sshPassword": {
+      "type": "securestring",
+      "metadata": {
+        "description": "The password must be at least 10 characters in length and must contain at least one digit, one non-alphanumeric character, and one upper or lower case letter."
+      }
+    },
+    "diskEncryptionVaultUri": {
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.json]: add a metadata description to all the parameters below

File: 101-hdinsight-spark-cmk/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+              "isDefault": true,
+              "container": "[parameters('clusterName')]",
+              "key": "[listKeys(resourceId('Microsoft.Storage/storageAccounts', variables('defaultStorageAccount').name), '2016-01-01').keys[0].value]"
+            }
+          ]
+        },
+        "computeProfile": {
+          "roles": [
+            {
+              "autoscale": null,
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.json]: remove all null properties

File: 101-hdinsight-spark-cmk/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+              "restAuthCredential.isEnabled": true,
+              "restAuthCredential.username": "[parameters('clusterLoginUserName')]",
+              "restAuthCredential.password": "[parameters('clusterLoginPassword')]"
+            }
+          }
+        },
+        "storageProfile": {
+          "storageaccounts": [
+            {
+              "name": "[replace(replace(reference(resourceId('Microsoft.Storage/storageAccounts', variables('defaultStorageAccount').name), '2016-01-01').primaryEndpoints.blob,'https://',''),'/','')]",
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.json]: update apiVersion

File: 101-hdinsight-spark-cmk/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+              "isDefault": true,
+              "container": "[parameters('clusterName')]",
+              "key": "[listKeys(resourceId('Microsoft.Storage/storageAccounts', variables('defaultStorageAccount').name), '2016-01-01').keys[0].value]"
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.json]: update apiVersion

File: 101-hdinsight-spark-cmk/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        },
+        "computeProfile": {
+          "roles": [
+            {
+              "autoscale": null,
+              "name": "headnode",
+              "minInstanceCount": 1,
+              "targetInstanceCount": 2,
+              "hardwareProfile": {
+                "vmSize": "Standard_D12_V2"
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.json]: make this a parameter

File: 101-hdinsight-spark-cmk/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+              },
+              "virtualNetworkProfile": null,
+              "dataDisksGroups": null
+            },
+            {
+              "autoscale": null,
+              "name": "workernode",
+              "targetInstanceCount": "[parameters('clusterWorkerNodeCount')]",
+              "hardwareProfile": {
+                "vmSize": "Standard_D13_V2"
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.json]: make this a parameter

File: 101-hdinsight-spark-cmk/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+              "dataDisksGroups": null
+            }
+          ]
+        },
+        "minSupportedTlsVersion": "[parameters('minTlsVersionNumber')]",
+        "diskEncryptionProperties": {
+          "vaultUri": "[parameters('diskEncryptionVaultUri')]",
+          "keyName": "[parameters('diskEncryptionKeyName')]",
+          "keyVersion": "[parameters('diskEncryptionKeyVersion')]",
+          "msiResourceId": "[resourceID('Microsoft.ManagedIdentity/userAssignedIdentities/', parameters('existingManagedIdentityName'))]"
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.json]: this needs to include the resourceGroup name paramter from the outputs

File: 101-hdinsight-spark-cmk/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+        }
+      },
+      "identity": {
+        "type": "UserAssigned",
+        "userAssignedIdentities": {
+          "[concat(resourceGroup().id, '/providers/Microsoft.ManagedIdentity/userAssignedIdentities/', parameters('existingManagedIdentityName'))]": {}
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.json]: remove concat, use resourceId and add the resourceGroup name parameter as above

File: 101-hdinsight-spark-cmk/azuredeploy.parameters.json [hunk 13]
```diff
@@ -0,0 +1,45 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+  "parameters": {
+    "clusterName": {
+      "value": "GEN-UNIQUE"
+    },
+    "clusterLoginUserName": {
+      "value": "admin"
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.parameters.json]: GEN-UNIQUE

File: 101-hdinsight-spark-cmk/azuredeploy.parameters.json [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    },
+    "clusterLoginPassword": {
+      "value": "GEN-PASSWORD"
+    },
+    "clusterWorkerNodeCount": {
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.parameters.json]: remove the parameters you have defaultValues for

File: 101-hdinsight-spark-cmk/azuredeploy.parameters.json [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+      "value": "3.6"
+    },
+    "sshUserName": {
+      "value": "sshuser"
+    },
+    "sshPassword": {
+      "value": "GEN-PASSWORD"
+    },
+    "diskEncryptionVaultUri": {
+      "value": "diskEncryptionVaultUri"
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.parameters.json]: GEN-KEYVAULT-ENCRYPTION-KEY-URI

File: 101-hdinsight-spark-cmk/azuredeploy.parameters.json [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    },
+    "diskEncryptionKeyName": {
+      "value": "diskEncryptionKeyName"
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.parameters.json]: GEN-KEYVAULT-ENCRYPTION-KEY

File: 101-hdinsight-spark-cmk/azuredeploy.parameters.json [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+    },
+    "diskEncryptionKeyVersion": {
+      "value": "diskEncryptionKeyVersion"
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.parameters.json]: GEN-KEYVAULT-ENCRYPTION-KEY-VERSION

File: 101-hdinsight-spark-cmk/azuredeploy.parameters.json [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    },
+    "minTlsVersionNumber": {
+      "value": "1.2"
+    },
+    "existingManagedIdentityName": {
```
Comment (bmoore-msft) [inline 101-hdinsight-spark-cmk/azuredeploy.parameters.json]: will need to add the resourceGroup name param for the MSI

**Comment by bmoore-msft** [comment]
@dagiro - you're going to need some change to get CI working due to the permission required in the subscription - don't worry about those perms for now.  Work on getting the comments address and some clear instructions in the README.md on what's needed for prereqs (given the trouble we've had, customers will also need some guidance).  Once those changes are done, I can finish up the prereqs.

**Pull request closed without merging**